src/darwin/Framework/CHIP/MTRDeviceStorageBehaviorConfiguration.h

```diff
@@ -0,0 +1,98 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+#import <Matter/MTRDefines.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * Class that configures how MTRDevice objects persist its attributes to storage, so as to not
+ * overwhelm the underlying storage system.
+ */
+MTR_NEWLY_AVAILABLE
+@interface MTRDeviceStorageBehaviorConfiguration : NSObject <NSCopying>
+
+/**
+ * Create configuration with a default set of values. See description below for details.
+ */
++ (instancetype)configurationWithDefaultStorageBehavior;
+
+/**
+ * Create configuration that disables storage behavior optimizations.
+ */
++ (instancetype)configurationWithStorageBehaviorOptimizationDisabled;
+
+/**
+ * Create configuration with specified values. See description below for details, and the list of
+ * properties below for valid ranges of these values.
+ */
++ (instancetype)configurationWithReportToPersistenceDelayTime:(NSTimeInterval)reportToPersistenceDelayTime
+                              reportToPersistenceDelayTimeMax:(NSTimeInterval)reportToPersistenceDelayTimeMax
+                                    recentReportTimesMaxCount:(NSUInteger)recentReportTimesMaxCount
+                          timeBetweenReportsTooShortThreshold:(NSTimeInterval)timeBetweenReportsTooShortThreshold
+                       timeBetweenReportsTooShortMinThreshold:(NSTimeInterval)timeBetweenReportsTooShortMinThreshold
+                        reportToPersistenceDelayMaxMultiplier:(double)reportToPersistenceDelayMaxMultiplier
+                  deviceReportingExcessivelyIntervalThreshold:(NSTimeInterval)deviceReportingExcessivelyIntervalThreshold;
+
+/**
+ * Storage behavior with values in the allowed range:
+ *
+ * Each time a report comes in, MTRDevice will wait reportToPersistDelayTime before persisting the
+ * changes to storage. If another report comes in during this internal, MTRDevice will wait another
+ * reportToPersistDelayTime interval, until reportToPersistDelayTimeMax is reached, at which
+ * point all the changes so far will be written to storage.
+ *
+ * MTRDevice will also track recentReportTimesMaxCount number of report times. If the running
+ * average time between reports dips below timeBetweenReportsTooShortThreshold, a portion of the
+ * reportToPersistenceDelayMaxMultiplier will be applied to both the reportToPersistenceDelayTime
+ * and reportToPersistenceDelayTimeMax. The multiplier will reach the max when the average time
+ * between reports reach timeBetweenReportsTooShortMinThreshold.
+ *
+ * When the running average time between reports dips below timeBetweenReportsTooShortMinThreshold
+ * for the first time, the time will be noted. If the device remains in this state for longer than
+ * deviceReportingExcessivelyIntervalThreshold, persistence will stop until the average time between
+ * reports go back above timeBetweenReportsTooShortMinThreshold.
```

Comment:
"goes back".

In terms of the implementation, if the device is reporting excessively for a bit, and then stops reporting entirely for a long time, shouldn't we persist at some point?  Right now, I don't think we do....

Seems to me like we should compute when the average time would go back above if we got no reports until then and schedule something to persist at that point or something.